### PR TITLE
Fixed scoring race condition

### PR DESCRIPTION
**Pull request by gsingh93**
There was a race condition which would allow people to score twice by submitting two requests concurrently. Teams could also request two hints even if they only had points for one hint, because of this race condition. This PR should fix both issues.

### CODE REVIEW

File: src/models/Level.php [hunk 1]
```diff
@@ -654,106 +654,156 @@ private static function levelFromRow(Map<string, string> $row): Level {
     );
   }
 
+  private static async function withLock(
+    string $lock_name,
+    int $team_id,
+    (function(): Awaitable<mixed>) $thunk,
+  ): Awaitable<mixed> {
+    $lock_name = sprintf('%s/%s_%d', sys_get_temp_dir(), $lock_name, $team_id);
```
Comment (javuto) [inline src/models/Level.php]: Do we want to provide timestamp here somewhere?

**Comment by javuto** [comment]
Looks pretty good to me! I added one comment just in case we want to make sure there is a isolation for the same team

File: src/models/Level.php (hunk 1 above)
Comment (gsingh93) [inline src/models/Level.php]: I was thinking of having each team reuse their lock file for the duration of the CTF. The benefit of this is we don't need to create a new file on each request. I don't see a downside of this reuse.

Comment (javuto) [inline src/models/Level.php]: Fair enough. That is a good solution. Push push! ![Approved] (http://a.fod4.com/misc/rambo%20original.gif)

**Pull request merged**